src/extract.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Recursively traverse the AST and return all variables and functions
+ * @param {object} node  
+ * @returns {object} an object with the list of variables and function
+ *  in the form {variables: [...], functions: [...]}
+ */
+export default function extract(node) {
+    switch(node.type) {
+        case "Literal":
+            return {variables: [], functions: []};
+        case "UnaryExpression":
+            return extract(node.argument);
+        case "BinaryExpression":
+            let right = extract(node.right);
+            // handles group(key: value)
+            if (node.operator === ":") {
+                return right;
+            }
```

Comment:
This special case is needed to handle Mavo's `group(key1: value1, key2: value2, ...)` operator. I'm thinking about the best way to be able to get rid of this special case by adding the second `options` parameter to the function. My first thought is to include the field `modifier` inside of options, i.e.

```ts
options: {
    modifier: (node) => void; // applies some sort of modification to the node
    // other options fields
}
```

This would then enable us (within Mavo) to call extract like this:

```js
const expression = "group(key1: 'foo', key2: 'bar')";
const ast = Mavo.Script.parse(expression);
const modifier = (node) => {
    if (node.type === "BinaryExpression" && node.operator === ":") {
        node = node.right;
    }
};
const extractResult = Vastly.extract(ast, {modifier})
```

Wanted to get feedback to see if anyone has any better thoughts on this, or the `options` parameter in general

Reply:
Indeed, Mavo specific stuff has no place here, but it's totally fine for Mavo to be passing specific options to these functions, so the general idea is on the right track: we design vastly to be flexible enough to accommodate Mavo, then Mavo can just pass certain options when using vastly methods.

Some issues with this proposed API:
- Only allowing a single function forces calling code to couple all their modifications in the same place
- The modifier being an arbitrary function will make it harder for you to adapt `extract()` to use this parameter. You can't add handling code to individual cases, so where does it go? Before or after the switch? 
- Speaking of ordering, it is also not clear to users when this gets executed: before, after, or instead of our code?
- I imagine a lot of use cases will want to specify these overrides once, rather than having to pass them in every single function call, so it would be nice if that was supported. As a rule of thumb, a good design principle for all UI design (including API design) is to try and make common things easy and complex things possible.

I have some thoughts on what an alternative API could look like, but I think there's value in you iterating on it before I share my thoughts. 😊